hail/python/hailtop/saige/saige.py

```diff
@@ -0,0 +1,388 @@
+                    null_glmm = self.config.step1_null_glmm(
+                        b,
+                        input_bfile=input_plink_data,
+                        input_phenotypes=input_phenotypes,
+                        phenotype=phenotype,
+                        analysis_type=SaigeAnalysisType.VARIANT,
+                        covariates=covariates,
+                        user_id_col=user_id_col,
+                        output_dir=checkpoint_dir.name,
+                        sparse_grm=sparse_grm,
+                    )
+                    for variant_chunk in variant_chunks:
+                        self.config.step2_spa(
+                            b,
+                            mt_path=mt_path,
+                            output_dir=output_dir,
+                            analysis_type=SaigeAnalysisType.VARIANT,
+                            null_model=null_glmm,
+                            input_data_type=input_data_type,
+                            chunk=variant_chunk,
+                            phenotype=phenotype,
+                            sparse_grm=sparse_grm,
+                            group_col=group_col,
+                        )
+
+            run_kwargs = run_kwargs or {}
+            run_kwargs.pop('wait')
+            b_handle = b.run(**run_kwargs, wait=False)
+
+            with hfs.open(f'{output_dir}/batch_info.yaml', 'w') as f:
+                info = yaml.safe_dump(
+                    {
+                        'batch_id': b_handle.id,
+                        'name': b.name,
+                        'attributes': b.attributes,
+                        'timestamp': str(datetime.datetime.now()),
+                    }
+                )
+                f.write(info + '\n')
+
+            b_handle.wait()
+
+            # FIXME: What do saige gene results tables look like?
+
+
+@app.command(help='run single variant version of SAIGE')
+def run_saige(
+    mt: Ann[str, Arg(help='Path to matrix table.')],
+    output_dir: Ann[str, Arg(help='Path to output directory for results.')],
+    phenotypes: Ann[
+        List[str], Arg(help='Phenotype names to run SAIGE on. Must be valid column fields in the matrix table.')
+    ],
+    covariates: Ann[
+        List[str], Arg(help='Covariates to use when running SAIGE. Must be valid column fields in the matrix table.')
+    ],
+    config: Ann[Optional[str], Opt(help='Path to SAIGE config file.')] = None,
+    checkpoint_dir: Ann[Optional[str], Opt(help='Path to output directory for checkpointing files.')] = None,
+    variant_chunks: Ann[
+        Optional[str], Opt(help='Path to variant interval file in either bed or interval list format.')
+    ] = None,
+    overrides: Ann[
+        Optional[List[str]],
+        Opt(help='Override configuration parameters in the default configuration. Example: sparse_grm.checkpoint=true'),
+    ] = None,
+):
+    if config is not None:
+        config = SaigeConfig.from_yaml_file(config)
```

Comment:
I'm a bit skeptical of YAML files for configuration. Why is this easier than using Python to configure it?

I think the YAML would be:
```
version=1
name='saige'
attributes:
    foo: bar
    ...
prepare_inputs:
    ...
...
```
And the Python would be:
```
from ... import run_saige

config = SaigeConfig(
    version=1,
    name='Saige',
    attributes={
        'foo': 'bar',
        ...
    },
    prepare_inputs=...
)

run_saige(config)
```

In the YAML case, I don't have an IDE that can give me hints about types and parameter names. If I get a parameter name wrong in YAML, I probably don't find out until we parse it? Or maybe it just gets ignored?

In Python, the IDE gives me hints and can give me red squiggles exactly where there's an error. Also, if I want to do some kind of preprocessing, automation, etc., I can just use Python to do that.

Reply:
I think it depends on how we want `hailtop.saige` (or similar) to be used. Do we want to support a command line interface or does the user have to directly use the Python functions and write their own wrappers and use the Python config classes? If we do support a command line interface, think it's a pain to have to use a bunch of command line options especially with all of the different SAIGE steps having their own parameters so I thought having a configuration file that was easy to specify the configuration would be easier. I was also inspired by the machine learning library I've been playing with where the configuration is done this way since there's just so many parameters to specify. Maybe it's overkill though for this use case. I can think about it some more.

Reply:
I agree a CLI is a nice way to invoke a simple tool. I'd rather to `cp x y` than `python3 -c 'import cp; cp("x", "y")'. However, once the interface is more complex than a CLI, the choice is a configuration file or general purpose programming language. If the choice is between those two, my preference is for the latter, but maybe I'm missing some clear win of the former. To be really specific it feels like the cli+config is:

```
./cli-tool --config config.yaml --maybe-another-option
```

And the python script is:

```
python3 config.py
```

Admittedly, if you want to pass extra command line options to config.py you have to use argparse or  `sys.argv`.